sabnzbd/dirscanner.py

```diff
@@ -115,99 +109,171 @@ def save(self):
                         continue
 
-                    # Add the NZB's
-                    res, _ = sabnzbd.nzbparser.add_nzbfile(path, catdir=catdir, keep=False)
-                    if res < 0:
-                        # Retry later, for example when we can't read the file
-                        self.suspected[path] = stat_tuple
-                    elif res == 0:
-                        self.error_reported = False
-                    else:
-                        self.ignored[path] = 1
+                    if path in self.suspected:
+                        if not compare_stat_tuple(self.suspected[path], stat_tuple):
+                            # Suspected file attributes have changed
+                            del self.suspected[path]
+
+                    yield path, catdir, stat_tuple
+        except OSError:
+            if not self.error_reported and not catdir:
+                logging.error(T("Cannot read Watched Folder %s"), filesystem.clip_path(folder))
+                self.error_reported = True
+
+    async def when_stable_add_nzbfile(self, path: str, catdir: Optional[str], stat_tuple: os.stat_result):
+        """Try and import the NZB but wait until the attributes are stable for 1 second, but give up after 3 sec"""
+
+        logging.info("Trying to import %s", path)
+
+        # Wait until the attributes are stable for 1 second, but give up after 3 sec
+        # This indicates that the file is fully written to disk
+        for n in range(3):
+            await asyncio.sleep(1.0)
+
+            try:
+                stat_tuple_tmp = os.stat(path)
+            except OSError:
+                continue
+            if compare_stat_tuple(stat_tuple, stat_tuple_tmp):
+                break
+            stat_tuple = stat_tuple_tmp
+        else:
+            # Not stable
+            return
+
+        # Add the NZB's
+        res, _ = sabnzbd.nzbparser.add_nzbfile(path, catdir=catdir, keep=False)
+        if res < 0:
+            # Retry later, for example when we can't read the file
+            self.suspected[path] = stat_tuple
+        elif res == 0:
+            self.error_reported = False
+        else:
+            self.ignored[path] = 1
+
+    def scan(self):
+        """Schedule a scan of the watched folder"""
+        if not self.loop:
+            return
+
+        if not (dirscan_dir := self.dirscan_dir):
+            return
+
+        asyncio.run_coroutine_threadsafe(self.scan_async(dirscan_dir), self.loop)
+
+    async def scan_async(self, dirscan_dir: str):
+        """Do one scan of the watched folder"""
+        async with self.lock:
+            if sabnzbd.PAUSED_ALL:
+                return
+
+            files: Set[str] = set()
+            futures: Set[asyncio.Task] = set()
+
+            for path, catdir, stat_tuple in self.get_suspected_files(dirscan_dir):
+                files.add(path)
+
+                if path in self.ignored or path in self.suspected:
+                    await asyncio.sleep(0)
```

Comment:
I'm not much versed in asyncio, so why is this 0-sleep needed?

Reply:
It would probably be fine without that one actually, but the one just after the create_task is yielding the event loop to give other coroutines a chance to run, the intention is that that create_task we just started can start running straight away because the first thing it will do is asyncio.sleep for a second anyway.

Reply:
Right, but the parent-corotine is almost finished at that point (line below)? So it will give the other ones a change to run almost immediately? Or I'm reading it wrong?

Reply:
My thought process was that the user could reasonably want to scan a directory with 20+ nzbs but the directory might have 100 items in it which we won't see but it still has to traverse the directory and maybe it could be slow to list everything?

It's probably a really quick operation even on HDDs so potentially remove both?